python_etl/churn.py

```diff
@@ -0,0 +1,654 @@
+        .withColumn('unique_domains_count_per_profile',
+                    average_udf(F.col('unique_domains_count'),
+                                F.col('n_profiles')))
+        # This value is meaningless because of overlapping domains between
+        # profiles
+        .drop('unique_domains_count')
+    )
+
+    return records_df
+
+
+def daterange(start_date, end_date):
+    for n in range(int((end_date - start_date).days) // 7):
+        yield (start_date + timedelta(n * 7)).strftime("%Y%m%d")
+
+
+def process_backfill(start_date, end_date, callback):
+    """ Import data from a start date to an end date.
+
+    :start_date ds: starting date in yyyymmdd
+    :end_date ds: ending date in yyymmdd
+    :callback: a callback accepting a datestring in format yyyymmdd
+    """
+    start_date = datetime.strptime(start_date, "%Y%m%d")
+    end_date = datetime.strptime(end_date, "%Y%m%d")
+    for d in daterange(start_date, end_date):
+        try:
+            callback(d)
+        except Exception as e:
+            print e
+
+
+def process_week(df, week_start, bucket, prefix):
+    result_df = compute_churn_week(df, week_start)
+
+    # Write to s3 as parquet, file size is on the order of 40MB. We
+    # bump the version number because v1 is the path to the old
+    # telemetry-batch-view churn data.
+    s3_path = "s3://{}/{}/week_start={}".format(bucket, prefix, week_start)
+    print("{}: Writing output as parquet to {}"
+          .format(datetime.utcnow(), s3_path))
+
+    result_df.repartition(1).write.parquet(s3_path, mode="overwrite")
+
+    print("Finished week {} at {}".format(week_start, datetime.utcnow()))
+
+
+@click.command()
+@click.option('-d', '--date', 'start_date',
+              default=fmt(datetime.now() - timedelta(7)),
+              help="datestring in form YYYYmmdd")
+@click.option('-b', '--bucket',
+              default='net-mozaws-prod-us-west-2-pipeline-analysis',
+              help="output s3 bucket")
+@click.option('-p', '--prefix', default='telemetry-test-bucket/churn-test',
+              help="output prefix associated with the s3 bucket")
+@click.option('--lag/--no-lag', default=True,
+              help="account for the 10 day collection period")
+@click.option('--backfill/--no-backfill', default=False,
+              help="backfill from the start date")
+def main(start_date, bucket, prefix, lag, backfill):
+    """Compute churn / retention information for unique segments of Firefox
+users acquired during a specific period of time.
+    """
+    spark = (SparkSession
+             .builder
+             .appName("churn")
+             .getOrCreate())
+
+    version = 'v2'
+    s3_path = "s3://{}/{}/{}".format(bucket, prefix, version)
+
+    # If this job is scheduled, we need the input date to lag a total of
+    # 10 days of slack for incoming data. Airflow subtracts 7 days to
+    # account for the weekly nature of this report.
+    week_start_date = start_date
+    if lag:
```

Comment:
Please factor out a function for computing lag and date-snapping and add some test coverage for different scenarios.